### DISCUSSION TITLE
Trying to avoid writing too many returns in if/else or match.

### DISCUSSION THREAD

**Discussion by jorgeluismireles** [vlang/v#22565] [Questions and Answers] [answered]
The combination of objects with options and created within a function with a **single** `return` outside cases `if/else` or `match` causes problems. The program:
```v
struct Number {
	natural ?string
	integer string
}

fn number(n int) Number {
	return if n >= 0 { // a single return here
		Number{ natural:'${n}', integer:'${n}' }
	} else {
		Number{ integer:'${n}' }
	}
}

fn main() {
	for n in [1,0,-1] {
		println('${number(n)}')
	}
}
```

https://play.vlang.io/p/78094800a4

The problem:
```
Running code...
Can't run code. The server returned an error:
/tmp/v_60000/../../../../../../box/code.v:7: error: ')' expected (got "_t2")
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
Exited with error status 1
Please try again.

```
A workaround is to use a `return` for **each case** (apart changing the `?string` by `string`):
```v
fn number(n int) Number {
	if n >= 0 {
		return Number{ natural:'${n}', integer:'${n}' } // a return here
	} else {
		return Number{ integer:'${n}' } // another return here
	}
}
```
The expected output:
```
Number{
    natural: Option('1')
    integer: '1'
}
Number{
    natural: Option('0')
    integer: '0'
}
Number{
    natural: Option(none)
    integer: '-1'
}
```

I have found this problem more than once trying to avoid writing "too many" `return`s in functions with a returning `match` with too many cases and the when the returned object is not *trivial*.

But now I wonder if is worth to solve the *issue* mentioned above for this particular case and then later more complex objects could cause new problems, or just encourage to program "the easy way" as in the workaround.

Thoughts?

**Comment by felipensp** [comment, marked as answer]
PR added to fix this issue #22578

**Reply by jorgeluismireles** [reply]
Great! Now https://play.vlang.io/p/78094800a4 works and `match` (which failed as `if` did) now works also:
```v
fn number(n int) Number {
	return match true {
		n >= 0 {
			Number{ natural:'${n}', integer:'${n}' }
		} else {
			Number{ integer:'${n}' }
		}
	}
}
```

**Reply by Delta456** [reply]
I will close this discussion. Thanks @felipensp
